scripts/asm-installer/install_asm

```diff
@@ -488,41 +488,74 @@ can_modify_at_all() {
 }
 
 can_modify_cluster_roles() {
-  if [[ "${ENABLE_ALL}" -eq 0 && "${ENABLE_CLUSTER_ROLES}" -eq 0 ]] && \
-    ! is_managed || \
-    ! can_modify_at_all; then false; fi
+  if ! can_modify_at_all; then false; return; fi
+
+  if is_managed || [[ "${ENABLE_ALL}" -eq 1 || "${ENABLE_CLUSTER_ROLES}" -eq 1 ]]; then
+    true
+  else
+    false
+  fi
 }
 
 can_modify_cluster_labels() {
-  if [[ "${ENABLE_ALL}" -eq 0 && "${ENABLE_CLUSTER_LABELS}" -eq 0 ]] \
-    || ! can_modify_at_all; then false; fi
+  if ! can_modify_at_all; then false; return; fi
+
+  if [[ "${ENABLE_ALL}" -eq 1 || "${ENABLE_CLUSTER_LABELS}" -eq 1 ]]; then
+    true
+  else
+    false
+  fi
 }
 
 can_modify_gcp_apis() {
-  if [[ "${ENABLE_ALL}" -eq 0 && "${ENABLE_GCP_APIS}" -eq 0 ]] \
-    || ! can_modify_at_all; then false; fi
+  if ! can_modify_at_all; then false; return; fi
+
+  if [[ "${ENABLE_ALL}" -eq 1 || "${ENABLE_GCP_APIS}" -eq 1 ]]; then
+    true
+  else
+    false
+  fi
 }
 
 can_modify_gcp_iam_roles() {
-  if [[ "${ENABLE_ALL}" -eq 0 && "${ENABLE_GCP_IAM_ROLES}" -eq 0 ]] && \
-    ! is_managed ||
-    ! can_modify_at_all; then false; fi
+  if ! can_modify_at_all; then false; return; fi
+
+  if is_managed || [[ "${ENABLE_ALL}" -eq 1 || "${ENABLE_GCP_IAM_ROLES}" -eq 1 ]]; then
+    true
+  else
+    false
+  fi
 }
 
 can_modify_gcp_components() {
-  if [[ "${ENABLE_ALL}" -eq 0 && "${ENABLE_GCP_COMPONENTS}" -eq 0 ]] && \
-    ! is_managed || \
-    ! can_modify_at_all; then false; fi
+  if ! can_modify_at_all; then false; return; fi
+
+  if is_managed || [[ "${ENABLE_ALL}" -eq 1 || "${ENABLE_GCP_COMPONENTS}" -eq 1 ]]; then
+    true
+  else
+    false
+  fi
 }
 
 can_register_cluster() {
-  if [[ "${ENABLE_ALL}" -eq 0 && "${ENABLE_REGISTRATION}" -eq 0 ]] \
-    || ! can_modify_at_all; then false; fi
+  if ! can_modify_at_all; then false; return; fi
+
+  if [[ "${ENABLE_ALL}" -eq 1 && ("${USE_VM}" -eq 1 || "${USE_HUB_WIP}" -eq 1) ]] \
```

Comment:
USE_HUB_WIP should be enough here. 

Im okay with this approach. Like previous release changes, we should also change the description of --enable-all to reflect this.

Reply:
USE_VM is there because HUB_WIP isn't the only environ dependency.

Reply:
We fail early if VM is not specified with hub-meshca. https://github.com/GoogleCloudPlatform/anthos-service-mesh-packages/blob/master/scripts/asm-installer/install_asm#L960

Reply:
I'm aware. Even if VMs didn't need HUB_WIP, it would still need environ registration. Removing it serves no purpose, and leaving it in makes that dependency more explicit.

Reply:
Good point. LGTM.